src/services/csv-generator.service.ts

```diff
@@ -0,0 +1,104 @@
+﻿import { Injectable } from "@nestjs/common";
+import { IoTDevice } from "@entities/iot-device.entity";
+import { IoTDeviceType } from "@enum/device-type.enum";
+import { Point } from "geojson";
+import { AuthenticationType } from "@enum/authentication-type.enum";
+import { forEach } from "lodash";
+import { EncryptionHelperService } from "@services/encryption-helper.service";
+
+@Injectable()
+export class CsvGeneratorService {
+    constructor(private encryptionHelperService: EncryptionHelperService) {}
+    public generateDeviceMetadataCsv(devices: IoTDevice[]) {
+        const csvHeader = csvFields.join(",");
+        let csvString = csvHeader + "\n";
+        devices.forEach(device => {
+            csvString +=
+                this.generateCsvRow((device as unknown) as IotDeviceCsvDto) + "\n";
+        });
+        return csvString;
+    }
+
+    // Having device as any required due to fields originating from different types
+    private generateCsvRow(device: any): string {
+        const {
+            name,
+            type,
+            location,
+            apiKey,
+            deviceEUI,
+            downlinkPayload,
+            mqttURL,
+            mqttPort,
+            mqtttopic,
+            authenticationType,
+            mqttusername,
+            caCertificate,
+            deviceCertificate,
+        } = device;
+
+        const mqttpassword = this.encryptionHelperService.basicDecrypt(
```

Comment:
This will not work since we can't decrypt a hashed password :-)

Reply:
Now only decrypting for MQTTSubscribers